File: app/src/main/java/com/example/caglarfindikli/springforandroidexample/MainActivity.java
package com.example.caglarfindikli.springforandroidexample;

import android.content.Intent;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.ColorMatrixColorFilter;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.view.Gravity;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.sdsmdg.tastytoast.TastyToast;
import com.squareup.picasso.Picasso;

import org.springframework.http.converter.StringHttpMessageConverter;
import org.springframework.web.client.RestTemplate;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

import nl.dionsegijn.konfetti.KonfettiView;
import nl.dionsegijn.konfetti.models.Shape;
import nl.dionsegijn.konfetti.models.Size;

import static com.example.caglarfindikli.springforandroidexample.R.id.imageView;
import static com.example.caglarfindikli.springforandroidexample.R.id.imageView5;


public class MainActivity extends AppCompatActivity {
    private static final Random RANDOM = new Random();
    private final int numberOfRoll = 1;
    private int sum1 = 0;
    private int sum2 = 0;
    private int[] tieBreakRoll = new int[2];
    private int counter1 = 0;
    private int counter2 = 0;
    private int bonusPoints1 = 0;
    private int bonusPoints2 = 0;
    private int currentDiceRollFirstCountry = 0;
    private int currentDiceRollSecondCountry = 0;
    private Button button;
    private TextView remainingRoll;
    private KonfettiView konfettiView1;


    private static int randomDiceValue() {
        return RANDOM.nextInt(6) + 1;
    }

    private static int randomCountry() {
        return RANDOM.nextInt(250) + 1;
    }

    private static String getKeyFromValue(Map hm, Object value) {
        for (Object o : hm.keySet()) {
            if (hm.get(o).equals(value)) {
                return o.toString().toLowerCase();
            }
        }
        return null;
    }

    private static android.net.Uri getFlag(String shortCode) {

        final String uri = "http://flagpedia.net/data/flags/normal/" + shortCode + ".png";

        return Uri.parse(uri);
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.animation_activty);
        final TextView firstCountry =
                findViewById(R.id.textView10);
        final TextView secondCountry = findViewById(R.id.textView11);
        final TextView firstCountryResult = findViewById(R.id.textView2);
        final TextView secondCountryResult = findViewById(R.id.textView4);
        remainingRoll = findViewById(R.id.textView3);
        remainingRoll.setText("Kalan Atış: " + numberOfRoll);
        final ImageView singleRollDiceResultFirstCountry = findViewById(R.id.imageView1);
        final ImageView singleRollDiceResultSecondCountry = findViewById(R.id.imageView2);
        final ImageView firstCountryFlag = findViewById(imageView);
        final ImageView secondCountryFlag = findViewById(imageView5);
        konfettiView1 = findViewById(R.id.konfettiView);
        final MediaPlayer mp = MediaPlayer.create(this, R.raw.queenwearethechampions);
        final MediaPlayer mp1 = MediaPlayer.create(this, R.raw.dicerolleffect);
        final MediaPlayer mp2 = MediaPlayer.create(this, R.raw.whawha);


        button = findViewById(R.id.rollDices);


        button.setOnClickListener(new View.OnClickListener() {
                                      public void onClick(View v) {

                                          final Animation anim1 = AnimationUtils.loadAnimation(MainActivity.this, R.anim.shake);
                                          final Animation anim2 = AnimationUtils.loadAnimation(MainActivity.this, R.anim.shake);
                                          final Animation.AnimationListener animationListener = new Animation.AnimationListener() {
                                              @Override
                                              public void onAnimationStart(Animation animation) {
                                                  mp1.start();

                                              }

                                              @Override
                                              public void onAnimationEnd(Animation animation) {
                                                  currentDiceRollFirstCountry = randomDiceValue();
                                                  currentDiceRollSecondCountry = randomDiceValue();

                                                  int res = getResources().getIdentifier("dice_" + currentDiceRollFirstCountry, "drawable", "com.example.caglarfindikli.springforandroidexample");
                                                  int res1 = getResources().getIdentifier("dice_" + currentDiceRollSecondCountry, "drawable", "com.example.caglarfindikli.springforandroidexample");

                                                  if (animation == anim1) {
                                                      if (counter1 < numberOfRoll) {
                                                          singleRollDiceResultFirstCountry.setImageResource(res);

                                                          sum1 += currentDiceRollFirstCountry;
                                                          firstCountryResult.setText(String.valueOf(sum1));
                                                          counter1++;
                                                          if (currentDiceRollFirstCountry == 6) {
                                                              bonusPoints1++;
                                                          }
                                                          if (currentDiceRollFirstCountry == 1) {
                                                              bonusPoints1--;
                                                          }
                                                      }


                                                  } else if (animation == anim2) {
                                                      if (counter2 < numberOfRoll) {
                                                          singleRollDiceResultSecondCountry.setImageResource(res1);
                                                          sum2 += currentDiceRollSecondCountry;
                                                          secondCountryResult.setText(String.valueOf(sum2));
                                                          counter2++;
                                                          if (currentDiceRollSecondCountry == 6) {
                                                              bonusPoints2++;
                                                          }
                                                          if (currentDiceRollSecondCountry == 1) {
                                                              bonusPoints2--;
                                                          }
                                                      }
                                                      if (numberOfRoll - counter1 == 0) {

                                                          if (sum1 > sum2) {

                                                              TastyToast.makeText(getApplicationContext(), "KAZANDIN!", TastyToast.LENGTH_LONG, TastyToast.SUCCESS).setGravity(Gravity.CENTER_VERTICAL | Gravity.CENTER_HORIZONTAL, 0, 0);

                                                              afterMatch(secondCountryFlag);
                                                              throwKonfetti(konfettiView1);
                                                              mp.start();
                                                              new Handler().postDelayed(new Runnable() {
                                                                  @Override
                                                                  public void run() {
                                                                      endGame();
                                                                      mp.stop();

                                                                  }
                                                              }, 8000);
                                                          } else if (sum2 > sum1) {

                                                              TastyToast.makeText(getApplicationContext(), "İKİ ZARA 80LİK OLDUN!", TastyToast.LENGTH_LONG, TastyToast.ERROR).setGravity(Gravity.CENTER_VERTICAL | Gravity.CENTER_HORIZONTAL, 0, 0);
                                                              afterMatch(firstCountryFlag);
                                                              setBW(firstCountryFlag);
                                                              mp2.start();

                                                              new Handler().postDelayed(new Runnable() {
                                                                  @Override
                                                                  public void run() {
                                                                      endGame();
                                                                      mp2.stop();
                                                                  }
                                                              }, 4000);
                                                          } else {
                                                              if (bonusPoints1 > bonusPoints2) {

                                                                  TastyToast.makeText(getApplicationContext(), "KAZANDIN\n" + "Avg: " + "(" + bonusPoints1 + ")" + "-" + "(" + bonusPoints2 + ")", TastyToast.LENGTH_LONG, TastyToast.SUCCESS).setGravity(Gravity.CENTER_VERTICAL | Gravity.CENTER_HORIZONTAL, 0, 0);
                                                                  afterMatch(secondCountryFlag);
                                                                  throwKonfetti(konfettiView1);
                                                                  mp.start();
                                                                  new Handler().postDelayed(new Runnable() {
                                                                      @Override
                                                                      public void run() {
                                                                          endGame();
                                                                          mp.stop();
                                                                      }
                                                                  }, 8000);
                                                              } else if (bonusPoints2 > bonusPoints1) {


                                                                  TastyToast.makeText(MainActivity.this, "İKİ ZARA 80LİK OLDUN\n" + "Avg: " + "(" + bonusPoints1 + ")" + "-" + "(" + bonusPoints2 + ")", TastyToast.LENGTH_LONG, TastyToast.ERROR).setGravity(Gravity.CENTER_VERTICAL | Gravity.CENTER_HORIZONTAL, 0, 0);
                                                                  afterMatch(firstCountryFlag);
                                                                  mp2.start();
                                                                  new Handler().postDelayed(new Runnable() {
                                                                      @Override
                                                                      public void run() {
                                                                          MainActivity.this.recreate();
                                                                      }
                                                                  }, 4000);
                                                              } else {
                                                                  tieBreakRoll[0] = randomDiceValue();
                                                                  tieBreakRoll[1] = randomDiceValue();
                                                                  if (tieBreakRoll[0] == tieBreakRoll[1]) {

                                                                      TastyToast.makeText(MainActivity.this, "KAZANDIN\n" + "Avg: " + "(" + bonusPoints1 + ")" + "-" + "(" + bonusPoints2 + ")" + " TieBreak Atış: " + tieBreakRoll[0] + "-" + tieBreakRoll[1], TastyToast.LENGTH_LONG, TastyToast.SUCCESS).setGravity(Gravity.CENTER_VERTICAL | Gravity.CENTER_HORIZONTAL, 0, 0);
                                                                      afterMatch(secondCountryFlag);
                                                                      throwKonfetti(konfettiView1);
                                                                      mp.start();
                                                                      new Handler().postDelayed(new Runnable() {
                                                                          @Override
                                                                          public void run() {
                                                                              endGame();
                                                                              mp.stop();
                                                                          }
                                                                      }, 8000);
                                                                  } else {

                                                                      TastyToast.makeText(MainActivity.this, "İKİ ZARA 80LİK OLDUN\n" + "Avg: " + "(" + bonusPoints1 + ")" + "-" + "(" + bonusPoints2 + ")" + " TieBreak Atış: " + tieBreakRoll[0] + "-" + tieBreakRoll[1], TastyToast.LENGTH_LONG, TastyToast.ERROR).setGravity(Gravity.CENTER_VERTICAL | Gravity.CENTER_HORIZONTAL, 0, 0);
                                                                      afterMatch(firstCountryFlag);
                                                                      mp2.start();
                                                                      new Handler().postDelayed(new Runnable() {
                                                                          @Override
                                                                          public void run() {
                                                                              endGame();
                                                                          }
                                                                      }, 4000);


                                                                  }

                                                              }

                                                          }
                                                      }


                                                  }
                                                  remainingRoll.setText("Kalan Atış: " + String.valueOf(numberOfRoll - counter1));
                                              }

                                              @Override
                                              public void onAnimationRepeat(Animation animation) {

                                              }
                                          };


                                          anim1.setAnimationListener(animationListener);
                                          anim2.setAnimationListener(animationListener);

                                          singleRollDiceResultFirstCountry.startAnimation(anim1);
                                          singleRollDiceResultSecondCountry.startAnimation(anim2);


                                      }


                                  }
        );


        final AsyncTask<String, Void, String> simpleGetTask =
                new AsyncTask<String, Void, String>() {
                    @Override
                    protected String doInBackground(String... params) {
                        RestTemplate restTemplate = new RestTemplate();
                        //add the String message converter, since the result of
                        // the call will be a String
                        restTemplate.getMessageConverters().add(
                                new StringHttpMessageConverter());
                        // Make the HTTP GET request on the url (params[0]),
                        // marshaling the response to a String
                        return
                                restTemplate.getForObject(params[0], String.class);
                    }

                    @Override
                    protected void onPostExecute(String result) {
                        // executed by the UI thread once the background
                        // thread is done getting the result
                        ObjectMapper mapper = new ObjectMapper();

                        Map<String, Object> map = new HashMap<String, Object>();

                        // convert JSON string to Map
                        try {
                            map = mapper.readValue(result, new TypeReference<Map<String, String>>() {
                            });
                        } catch (IOException e) {
                            e.printStackTrace();
                        }

                        ArrayList<Object> countries = new ArrayList<>(map.values());
                        int randomSelectionFirstCountry = randomCountry();
                        int randomSelectionSecondCountry = randomCountry();
                        try {
                            Thread.currentThread().sleep(2000);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                        firstCountry.setText(countries.get(randomSelectionFirstCountry - 1).toString());
                        Picasso.with(MainActivity.this).load(getFlag(getKeyFromValue(map, countries.get(randomSelectionFirstCountry - 1)))).resize(400, 267).error(R.drawable.rollingdices).into(firstCountryFlag);
                        firstCountryFlag.setImageURI(getFlag(getKeyFromValue(map, countries.get(randomSelectionFirstCountry - 1))));
                        secondCountry.setText(countries.get(randomSelectionSecondCountry - 1).toString());
                        Picasso.with(MainActivity.this).load(getFlag(getKeyFromValue(map, countries.get(randomSelectionSecondCountry - 1)))).resize(400, 267).error(R.drawable.rollingdices).into(secondCountryFlag);
                        secondCountryFlag.setImageURI(getFlag(getKeyFromValue(map, countries.get(randomSelectionSecondCountry - 1))));


                    }
                };

        String url = "http://country.io/names.json";
        simpleGetTask.execute(url);


    }


    private void setBW(ImageView iv) {

        float brightness = 10; // change values to suite your need

        float[] colorMatrix = {
                0.33f, 0.33f, 0.33f, 0, brightness,
                0.33f, 0.33f, 0.33f, 0, brightness,
                0.33f, 0.33f, 0.33f, 0, brightness,
                0, 0, 0, 1, 0
        };

        ColorFilter colorFilter = new ColorMatrixColorFilter(colorMatrix);
        iv.setColorFilter(colorFilter);
    }

    private void throwKonfetti(KonfettiView konfettiView) {
        konfettiView.build()
                .addColors(Color.YELLOW, Color.GREEN, Color.MAGENTA, Color.BLUE)
                .setDirection(0.0, 359.0)
                .setSpeed(1f, 5f)
                .setFadeOutEnabled(true)
                .setTimeToLive(2000L)
                .addShapes(Shape.RECT, Shape.CIRCLE)
                .addSizes(new Size(12, 5f))
                .setPosition(-50f, konfettiView.getWidth() + 50f, -50f, -50f)
                .streamFor(600, 5000L);
    }

    private void endGame() {

        Intent intent = new Intent(this, StartActivity.class);
        startActivity(intent);
    }

    private void afterMatch(ImageView imageView) {

        button.setVisibility(View.INVISIBLE);
        remainingRoll.setVisibility(View.INVISIBLE);
        setBW(imageView);


    }


}

File: app/src/main/java/com/cfindikli/apps/rollingCountries/View/MainActivity.kt
package com.cfindikli.apps.rollingCountries

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.hardware.SensorManager
import android.media.MediaPlayer
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.text.Html
import android.view.Gravity
import android.view.View
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.cfindikli.apps.rollingCountries.Model.CountryModel
import com.cfindikli.apps.rollingCountries.Utils.Companion.firstCountryObj
import com.cfindikli.apps.rollingCountries.Utils.Companion.secondCountryObj
import com.cfindikli.apps.rollingCountries.View.StartActivity
import com.sdsmdg.tastytoast.TastyToast
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import com.squareup.seismic.ShakeDetector
import kotlinx.android.synthetic.main.animation_activty.*
import java.io.IOException
import java.util.*


@Suppress("DEPRECATION")
class MainActivity : AppCompatActivity(), ShakeDetector.Listener {


    @SuppressLint("SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.animation_activty)
        adjustUIComponent()
        initializeMediaPlayer()
        setListeners()
        setUI()
        initializeShakeDetector()
    }


    private fun initializeMediaPlayer() {
        Utils.mp = MediaPlayer.create(applicationContext, R.raw.dicerolleffect)
    }


    @SuppressLint("SetTextI18n")
    internal fun adjustUIComponent() {
        levelName.text = firstCountryObj.levelName[firstCountryObj.level]
        remainingRoll!!.text = resources.getString(R.string.text_remaining_roll) + firstCountryObj.numberOfRoll
        resetValues()
        singleRollDiceResultFirstCountry!!.visibility = View.INVISIBLE
        singleRollDiceResultSecondCountry!!.visibility = View.INVISIBLE
    }


    private fun initializeShakeDetector() {
        Utils.sensorManager = getSystemService(Context.SENSOR_SERVICE) as SensorManager
        Utils.shakeDetector = ShakeDetector(this)
        Utils.shakeDetector!!.start(Utils.sensorManager)
    }

    private fun setListeners() {
        rollDiceButton!!.setOnClickListener(ClickListener())
        volumeIcon!!.setOnClickListener(ClickListener())
    }


    internal fun volumeOnOff() {
        Utils.isMute = !Utils.isMute
        when {
            Utils.isMute -> {
                volumeIcon!!.setImageResource(R.drawable.mute)
                Utils.mp!!.setVolume(0f, 0f)
            }
            else -> {
                Utils.isMute = false
                volumeIcon!!.setImageResource(R.drawable.volume)
                Utils.mp!!.setVolume(1f, 1f)
            }
        }
    }

    internal fun rollDice() {
        Utils.anim1 = AnimationUtils.loadAnimation(this@MainActivity, R.anim.shake)
        Utils.anim2 = AnimationUtils.loadAnimation(this@MainActivity, R.anim.shake)
        val animationListener = object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {
                rollDiceButton!!.visibility = View.INVISIBLE
                changeTrack(0)
                singleRollDiceResultFirstCountry!!.visibility = View.VISIBLE
                singleRollDiceResultSecondCountry!!.visibility = View.VISIBLE
            }

            @SuppressLint("SetTextI18n")
            override fun onAnimationEnd(animation: Animation) {
                firstCountryObj.currentDiceRoll = Utils.randomDiceValue()[0]
                secondCountryObj.currentDiceRoll = Utils.randomDiceValue()[1]
                val firstDice = resources.getIdentifier("dice_${firstCountryObj.currentDiceRoll}", "drawable", packageName)
                val secondDice = resources.getIdentifier("dice_${secondCountryObj.currentDiceRoll}", "drawable", packageName)


                when {
                    firstCountryObj.numberOfRoll != 0 -> {

                        when (animation) {
                            Utils.anim1 -> {
                                singleRollDiceResultFirstCountry!!.setImageResource(firstDice)
                                firstCountryObj.sum += firstCountryObj.currentDiceRoll
                                firstCountryResult.text = firstCountryObj.sum.toString()
                                singleRollDiceResultSecondCountry!!.setImageResource(secondDice)
                                secondCountryObj.sum += secondCountryObj.currentDiceRoll
                                secondCountryResult.text = secondCountryObj.sum.toString()
                                firstCountryObj.numberOfRoll--
                                remainingRoll!!.text = resources.getString(R.string.text_remaining_roll) + firstCountryObj.numberOfRoll.toString()
                                assesEarlyWinning()
                            }

                        }
                    }
                    else -> {
                        assessResult()
                    }
                }

            }

            override fun onAnimationRepeat(animation: Animation) {

            }
        }

        Utils.anim1!!.setAnimationListener(animationListener)
        Utils.anim2!!.setAnimationListener(animationListener)

        singleRollDiceResultFirstCountry!!.startAnimation(Utils.anim1)
        singleRollDiceResultSecondCountry!!.startAnimation(Utils.anim2)

    }

    private fun setUI() {
        Picasso.get().load(firstCountryObj.imageUrl).resize(400, 267)
                .into(firstCountryFlag!!, object : Callback {
                    override fun onSuccess() {
                        firstCountry.text = firstCountryObj.name
                    }

                    override fun onError(e: Exception?) {
                        firstCountry.text = firstCountryObj.name
                        firstCountryObj.flag = resources.getIdentifier("flag_" + firstCountryObj.alpha2Code, "drawable", packageName)
                        firstCountryFlag!!.setImageResource(firstCountryObj.flag!!)
                    }
                })

        Picasso.get().load(secondCountryObj.imageUrl).resize(400, 267)
                .into(secondCountryFlag!!, object : Callback {
                    override fun onSuccess() {
                        secondCountry.text = secondCountryObj.name
                    }

                    override fun onError(e: Exception?) {
                        secondCountry.text = secondCountryObj.name
                        secondCountryObj.flag = resources.getIdentifier("flag_" + secondCountryObj.alpha2Code, "drawable", packageName)
                        secondCountryFlag!!.setImageResource(secondCountryObj.flag!!)
                    }
                })
    }


    private fun changeTrack(position: Int) {
        Utils.mp!!.stop()
        Utils.mp!!.reset()
        try {
            Utils.mp!!.setDataSource(applicationContext, Uri.parse("android.resource://" + packageName + "/" + Utils.song[position]))
        } catch (e: IOException) {
            e.printStackTrace()
        }

        try {
            Utils.mp!!.prepare()
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: IllegalStateException) {
            e.printStackTrace()
        }

        Utils.mp!!.start()
    }

    private fun endGame(imageView: ImageView?) {
        Utils.mp!!.stop()
        Utils.setBW(imageView!!, 1f)
        showAfterMatchDialog()
    }

    @SuppressLint("SetTextI18n")
    private fun afterMatch(imageView: ImageView?) {
        rollDiceButton!!.visibility = View.INVISIBLE
        remainingRoll!!.visibility = View.INVISIBLE
        Utils.setBW(imageView!!, 0f)
    }


    @SuppressLint("ResourceType", "SetTextI18n")
    private fun rematch(level: Int, reselectType: Int) {
        when {

            level < firstCountryObj.levelName.size -> {

                when (reselectType) {
                    1 -> {
                        Utils.reselect(firstCountryObj)
                        goBackPreviousActivity()
                    }
                    2 -> {
                        Utils.reselect(secondCountryObj)
                        TastyToast.makeText(applicationContext, "Your Next Opponent: ${secondCountryObj.name}", TastyToast.LENGTH_SHORT, TastyToast.DEFAULT).setGravity(Gravity.CENTER_VERTICAL or Gravity.CENTER_HORIZONTAL, 0, 0)
                    }
                    3 -> {
                        Utils.reselect(firstCountryObj)
                        Utils.reselect(secondCountryObj)
                    }
                }
                setUI()
            }
        }
        resetValues()
        Utils.shakeDetector!!.start(Utils.sensorManager)
    }

    private fun assessResult() {
        Utils.shakeDetector!!.stop()
        when {
            firstCountryObj.sum > secondCountryObj.sum -> {
                ++firstCountryObj.level
                afterMatch(secondCountryFlag)
                winningCeremony(firstCountryObj.level)
            }
            secondCountryObj.sum > firstCountryObj.sum -> {
                firstCountryObj.level = 0
                TastyToast.makeText(applicationContext, "YOU LOSE!", TastyToast.LENGTH_LONG, TastyToast.ERROR).setGravity(Gravity.CENTER_VERTICAL or Gravity.CENTER_HORIZONTAL, 0, 0)
                afterMatch(firstCountryFlag)
                Utils.setBW(firstCountryFlag!!, 0f)
                changeTrack(2)
                Handler().postDelayed({ endGame(firstCountryFlag) }, 4000)
            }
            else -> {
                ++firstCountryObj.numberOfRoll
                rollDiceButton.text = resources.getString(R.string.text_tie_break)
                remainingRoll!!.visibility = View.INVISIBLE
            }
        }

    }

    private fun assesEarlyWinning() {

        when {
            (firstCountryObj.numberOfRoll == 1 && (firstCountryObj.sum - secondCountryObj.sum >= 6 || secondCountryObj.sum - firstCountryObj.sum >= 6)) -> {
                assessResult()
            }
            (firstCountryObj.numberOfRoll == 2 && (firstCountryObj.sum - secondCountryObj.sum >= 11 || secondCountryObj.sum - firstCountryObj.sum >= 11)) -> {
                assessResult()
            }
            else -> {
                rollDiceButton!!.visibility = View.VISIBLE
            }
        }
    }

    private fun winningCeremony(level: Int) {
        when (level) {
            firstCountryObj.levelName.size -> {
                Utils.throwConfetti(konfettiView!!)
                changeTrack(1)
                TastyToast.makeText(applicationContext, "THE WORLD CHAMPIONS\n" + firstCountryObj.name!!.toUpperCase(Locale.getDefault()), TastyToast.LENGTH_LONG, TastyToast.SUCCESS).setGravity(Gravity.CENTER_VERTICAL or Gravity.CENTER_HORIZONTAL, 0, 0)
                Handler().postDelayed({ endWinningCeremony(secondCountryFlag) }, 37000)
            }
            else -> {
                TastyToast.makeText(applicationContext, "YOU WIN!", TastyToast.LENGTH_LONG, TastyToast.SUCCESS).setGravity(Gravity.CENTER_VERTICAL or Gravity.CENTER_HORIZONTAL, 0, 0)
                Handler().postDelayed({ endWinningCeremony(secondCountryFlag) }, 4000)
            }
        }
    }


    private fun endWinningCeremony(imageView: ImageView?) {

        Utils.mp!!.stop()
        Utils.setBW(imageView!!, 1f)

        when {
            firstCountryObj.level < firstCountryObj.levelName.size -> {
                firstCountryObj.reselectType = 2
                rematch(firstCountryObj.level, firstCountryObj.reselectType)
            }
            else -> {
                showAfterMatchDialog()
            }
        }

    }

    private fun resetValues() {
        firstCountryObj.numberOfRoll = CountryModel().numberOfRoll
        firstCountryObj.sum = 0
        secondCountryObj.sum = 0
        levelName.text = firstCountryObj.levelName[firstCountryObj.level]
        rollDiceButton!!.visibility = View.VISIBLE
        rollDiceButton!!.text = resources.getString(R.string.text_roll_dice)
        remainingRoll!!.visibility = View.VISIBLE
        remainingRoll!!.text = resources.getString(R.string.text_remaining_roll).plus(firstCountryObj.numberOfRoll.toString())
        singleRollDiceResultFirstCountry!!.visibility = View.VISIBLE
        singleRollDiceResultSecondCountry!!.visibility = View.VISIBLE
        firstCountryResult.text = firstCountryObj.sum.toString()
        secondCountryResult.text = secondCountryObj.sum.toString()
        singleRollDiceResultFirstCountry!!.setImageResource(R.drawable.dice_6)
        singleRollDiceResultSecondCountry!!.setImageResource(R.drawable.dice_6)
        singleRollDiceResultFirstCountry!!.visibility = View.INVISIBLE
        singleRollDiceResultSecondCountry!!.visibility = View.INVISIBLE
    }

    private fun showAfterMatchDialog() {

        val alertDialog = AlertDialog.Builder(this@MainActivity)
        alertDialog.setMessage(R.string.after_match_text)
        alertDialog.setCancelable(false)

        alertDialog.setNeutralButton(Html.fromHtml("<font color='#3342FF'>Change Both Teams (Automatically)</font>")) { _, _ ->
            firstCountryObj.reselectType = 3
            rematch(firstCountryObj.level, firstCountryObj.reselectType)
        }

        alertDialog.setNegativeButton(Html.fromHtml("<font color='#3342FF'>Change Opponent (Automatically)</font>")) { dialog, _ ->
            firstCountryObj.reselectType = 2
            rematch(firstCountryObj.level, firstCountryObj.reselectType)
            dialog.cancel()
        }

        alertDialog.setPositiveButton(Html.fromHtml("<font color='#3342FF'>Just Change My Team (Manually)</font>")) { dialog, _ ->
            firstCountryObj.reselectType = 1
            rematch(firstCountryObj.level, firstCountryObj.reselectType)
            dialog.cancel()
        }


        val alert = alertDialog.create()
        alert.show()

        val mw = alert.findViewById<TextView>(android.R.id.message)
        mw.gravity = Gravity.CENTER

    }

    private fun showQuitDialog() {

        val alertDialog = AlertDialog.Builder(this@MainActivity)
        alertDialog.setMessage(R.string.exit_game_text)
        alertDialog.setCancelable(false)

        alertDialog.setPositiveButton(Html.fromHtml("<font color='#3342FF'>Yes</font>")) { _, _ ->
            finishAffinity()

        }

        alertDialog.setNegativeButton(Html.fromHtml("<font color='#3342FF'>No</font>")) { dialog, _ ->
            dialog.cancel()
        }

        val alert = alertDialog.create()
        alert.show()

        val mw = alert.findViewById<TextView>(android.R.id.message)
        mw.gravity = Gravity.CENTER

    }

    private fun goBackPreviousActivity() {
        val i = Intent(this, StartActivity::class.java)
        startActivity(i)
    }

    override fun onBackPressed() {
        showQuitDialog()
    }

    override fun hearShake() {
        rollDice()
    }

    internal inner class ClickListener : View.OnClickListener {
        override fun onClick(view: View) {
            when (view.id) {
                R.id.rollDiceButton -> rollDice()
                R.id.volumeIcon -> volumeOnOff()
            }
        }
    }


}









